### ISSUE TITLE
must_epoch+indexlaunch+fill = bad data

### ISSUE THREAD

**Issue by streichler** [StanfordLegion/legion#157]
The following test fails because the updates made by the index launch seem to get overwritten by a duplicate fill operation:
https://github.com/StanfordLegion/legion/blob/54a01af7dd16a9f3cc103030ec2c9b7c290d113e/language/tests/runtime/bugs/must_fill_index.rg

When run with legion spy, there is no error reported, but the legion spy dataflow graphs do show the multiple fills.

**Comment by lightsighter** [comment]
Fix pushed with e68b529.  Assigning back to Sean to confirm the fix.

**Comment by lightsighter** [comment]
Also worth noting that Legion Spy did report a pertinent warning.

**Comment by streichler** [comment]
What was the warning?  Perhaps I had the wrong set of command line options?

**Comment by lightsighter** [comment]
hellbender:language mebauer$ pypy ../tools/legion_spy.py -lpa spy.log 
Reading log file spy.log...
WARNING: Skipped 1 lines when reading spy.log
Reducing top-level index space shapes...
Done
Simplifying event graph...
Done
Performing logical analysis...
Performing logical dependence analysis for main...
Pass
Performing logical dependence analysis for k...
Pass
Performing physical analysis...
  Performing physical dependence analysis for k (UID 3)...
    Performing physical dependence analysis for Mapping Op 5 (UID 5)...
    Performing physical dependence analysis for Fill Op 6 (UID 6)...
    Performing physical dependence analysis for g (UID 14)...
    Performing physical dependence analysis for g (UID 15)...
    Performing physical dependence analysis for g (UID 16)...
    Performing physical dependence analysis for Inter Close Op 10 generated by Mapping Op 9
    Performing physical dependence analysis for Mapping Op 9 (UID 9)...
**WARNING: Mapping Op 9 generated 1 unnecessary Realm fills**
    Realm Fill (4) was unnecessary
  Performing physical dependence analysis for __binary_int32_int32_ (UID 4)...
Pass
Legion Spy analysis complete.  Exiting...

**Comment by streichler** [comment]
Ah - the "Pass" at the end threw me off.  I'll probably go and modify legion_spy to report the number of errors/warnings as the last thing it does, and maybe reduce the spewage unless a "verbose" option is specified.

**Comment by streichler** [comment]
Also, what's the difference between "warning" and "error" here?  The fill was not just unnecessary, it was incorrect, resulting in bad data.

**Comment by lightsighter** [comment]
Yes, but there are some cases where the Legion Spy analysis is more precise than the runtime's analysis, and the result is that it can't tell when a fill/copy is just unnecessary or wrong.  With version number information it could tell the difference.

**Comment by streichler** [comment]
Also, while e68b529 fixes the cut-down Regent repro case, the original test case is still failing.  I'm looking at what needs to be added to the Regent test.

**Comment by streichler** [comment]
Commit b3ce9f7 has a new Regent test that mimics the failure mode of the original test case.  An index launch is performed in which each index task has simultaneous read/write access to the same regions.  The previously-performed fills on the regions are repeated between the index tasks, overwriting data written by the earlier tasks, and could cause deadlock in the must epoch case.

Legion spy reports the following errors, and shows dependencies between the index tasks, even in the must epoch case:
```    Performing physical dependence analysis for h (UID 17)...
WARNING: h generated 3 unnecessary Realm fills
    Realm Fill (4) was unnecessary
    Realm Fill (5) was unnecessary
    Realm Fill (6) was unnecessary
    Performing physical dependence analysis for h (UID 18)...
WARNING: h generated 3 unnecessary Realm fills
    Realm Fill (7) was unnecessary
    Realm Fill (8) was unnecessary
    Realm Fill (9) was unnecessary
    Performing physical dependence analysis for Inter Close Op 12 generated by Mapping Op 11
```

**Comment by lightsighter** [comment]
This should be fixed with 33dd491 and validates with 4bc2780.  Assigning back to @streichler to confirm the fix.

**Comment by streichler** [comment]
Yep, this looks good, both on the mini regent test case and the original app.
